plugins/ata/fu-ata-device.c

```diff
@@ -0,0 +1,612 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fcntl.h>
+#include <string.h>
+#include <sys/errno.h>
+#include <sys/ioctl.h>
+#include <scsi/sg.h>
+#include <glib/gstdio.h>
+
+#include "fu-ata-device.h"
+#include "fu-chunk.h"
+
+#define FU_ATA_IDENTIFY_SIZE	512	/* bytes */
+#define FU_ATA_BLOCK_SIZE	512	/* bytes */
+
+struct ata_tf {
+	guint8			dev;
+	guint8			command;
+	guint8			error;
+	guint8			status;
+	guint8			feat;
+	guint8			nsect;
+	guint8			lbal;
+	guint8			lbam;
+	guint8			lbah;
+};
+
+#define ATA_USING_LBA			(1 << 6)
+#define ATA_STAT_DRQ			(1 << 3)
+#define ATA_STAT_ERR			(1 << 0)
+
+#define ATA_OP_IDENTIFY			0xec
+#define ATA_OP_DOWNLOAD_MICROCODE	0x92
+
+#define SG_CHECK_CONDITION		0x02
+#define SG_DRIVER_SENSE			0x08
+
+#define SG_ATA_12			0xa1
+#define SG_ATA_12_LEN			12
+
+#define SG_ATA_PROTO_NON_DATA		(3 << 1)
+#define SG_ATA_PROTO_PIO_IN		(4 << 1)
+#define SG_ATA_PROTO_PIO_OUT		(5 << 1)
+
+enum {
+	SG_CDB2_TLEN_NODATA	= 0 << 0,
+	SG_CDB2_TLEN_FEAT	= 1 << 0,
+	SG_CDB2_TLEN_NSECT	= 2 << 0,
+
+	SG_CDB2_TLEN_BYTES	= 0 << 2,
+	SG_CDB2_TLEN_SECTORS	= 1 << 2,
+
+	SG_CDB2_TDIR_TO_DEV	= 0 << 3,
+	SG_CDB2_TDIR_FROM_DEV	= 1 << 3,
+
+	SG_CDB2_CHECK_COND	= 1 << 5,
+};
+
+struct _FuAtaDevice {
+	FuUdevDevice		 parent_instance;
+	guint			 pci_depth;
+	gint			 fd;
+	guint64			 transfer_size;
```

Comment:
This one can also only go up to `0xfff` right?  So it should probably be a `guint16`

Reply:
This is 0xffff * 512 surely?

Reply:
In readme you're saying that  the biggest it will take is `0xffff`.  So I guess there is confusion if it's supposed to be a multiple of ATA_BLOCK_SIZE or not when you provide in the quirk.  Since you don't multiple it in the quirk.

Reply:
I guess it is confusing. Do you think it should be the number of blocks or the number of bytes in the quirk file?

Reply:
Yeah that's fine, and much clearer.